lint-staged.config.js

```diff
@@ -0,0 +1,60 @@
+/**
+ * External dependencies
+ */
+const path = require( 'path' );
+const micromatch = require( 'micromatch' );
+
+/**
+ * Internal dependencies
+ */
+const { plugins } = require( './plugins.json' );
+
+/**
+ * Join and escape filenames for shell.
+ *
+ * @param {string[]} files Files to join.
+ *
+ * @return {string} Joined files.
+ */
+const joinFiles = ( files ) => {
+	return files.map( ( file ) => `'${ file }'` ).join( ' ' );
+};
+
+// Get plugin base name to match regex more accurately.
+// Else it can cause issues when this plugin is places in `wp-content/plugins` directory.
+const PLUGIN_BASE_NAME = path.basename( __dirname );
+
+module.exports = {
+	'**/*.js': ( files ) => `npm run lint-js ${ joinFiles( files ) }`,
+	'**/*.php': ( files ) => {
+		const commands = [ 'composer phpstan' ];
```

Comment:
As noted by @thelovekesh in https://github.com/WordPress/performance/pull/1228#issuecomment-2125658548, this intentionally does _not_ include the list of changed files as arguments as PHPStan should be run across the entire codebase.